hypothesis-python/src/hypothesis/ghostwriter.py

```diff
@@ -173,14 +251,191 @@ def _make_test(
+    """Write source code for a property-based test of ``func``.
+
+    The resulting test checks that if you call ``func`` on it's own output,
+    the result does not change.  For example:
+
+    .. code-block:: python
+
+        from typing import Sequence
+        from hypothesis import ghostwriter
+
+        def timsort(seq: Sequence[int]) -> Sequence[int]:
+            return sorted(seq)
+
+        ghostwriter.idempotent(timsort)
+
+    Gives:
+
+    .. code-block:: python
+
+        # This test template was produced by the `hypothesis.ghostwriter` module.
+        import typing
+        from hypothesis import assume, given
+        from hypothesis.strategies import *
+
+        @given(seq=one_of(lists(elements=integers()), binary(), binary().map(bytearray)))
+        def test_idempotent_timsort(seq):
+            result = timsort(seq=sequence)
+            assert timsort(result) == result
+    """
+    if not callable(func):
+        raise InvalidArgument("Got non-callable func=%r" % (func,))
+    except_ = _check_except(except_)
+    _check_style(style)
+
+    params = _get_params(func)
+    test_body = (
+        "result = {0}.{1}({2})\n" + _assert_eq(style, "{0}.{1}(result)", "result")
+    ).format(
+        func.__module__, func.__name__, ", ".join("{0}={0}".format(p) for p in params)
+    )
+    return _make_test(
+        func, test_body=test_body, except_=except_, ghost="idempotent", style=style
+    )
+
+
+def roundtrip(*funcs: Callable, except_: Except = (), style: str = "pytest") -> str:
+    """Write source code for a property-based test of ``funcs``.
+
+    The resulting test checks that if you call the first function, pass the result
+    to the second (and so on), the final result is equal to the first input argument.
+
+    This is a *very* powerful property to test, especially when the config options
+    are varied along with the object to round-trip.  For example, try ghostwriting
+    a test for :func:`python:json.dumps` - would you have thought of all that?
+
+    .. code-block:: shell
+
+        hypothesis write roundtrip json.dumps json.loads
+    """
+    if not funcs:
+        raise InvalidArgument("Round-trip of zero functions is meaningless.")
+    for i, f in enumerate(funcs):
+        if not callable(f):
+            raise InvalidArgument("Got non-callable funcs[%s]=%r" % (i, f))
+    except_ = _check_except(except_)
+    _check_style(style)
+
+    test_body = ""
+    for i, f in enumerate(funcs):
+        params = list(_get_params(f))
+        if i == 0:
+            args = ", ".join("{0}={0}".format(p) for p in params)
+        else:
+            args = ", ".join(
+                "{}={}".format(k, v)
+                for k, v in [
+                    (params[0], "value{}".format(i - 1)),
+                    *zip(params[1:], params[1:]),
+                ]
+            )
+        test_body += "value{} = {}.{}({})\n".format(i, f.__module__, f.__name__, args)
```

Comment:
Yeah I think I actually really want this code to use the pretty printer module or something like it to write its results. This kind of string concatenation code is fairly unpleasant to read.